client-java/src/main/java/uk/gov/gchq/palisade/client/internal/resource/WebSocketClient.java

```diff
@@ -0,0 +1,191 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.palisade.client.internal.resource;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import uk.gov.gchq.palisade.client.internal.resource.WebSocketListener.Item;
+import uk.gov.gchq.palisade.client.util.Checks;
+import uk.gov.gchq.palisade.client.util.ImmutableStyle;
+
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.WebSocket;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.function.UnaryOperator;
+
+import static uk.gov.gchq.palisade.client.internal.resource.WebSocketListener.createResourceClientListener;
+
+/**
+ * An instance of this class manages the communications to the Filtered Resource
+ * Service via WebSockets
+ *
+ * @since 0.5.0
+ */
+public class WebSocketClient {
+
+    /**
+     * Provides service and configuration for a resource client
+     *
+     * @since 0.5.0
+     */
+    @Value.Immutable
+    @ImmutableStyle
+    public interface ResourceClientSetup {
+
+        /**
+         * Exposes the generated builder outside this package
+         * <p>
+         * While the generated implementation (and consequently its builder) is not
+         * visible outside of this package. This builder inherits and exposes all public
+         * methods defined on the generated implementation's Builder class.
+         */
+        class Builder extends ImmutableResourceClientSetup.Builder { // empty
+        }
+
+        /**
+         * Returns the token
+         *
+         * @return the token
+         */
+        String getToken();
+
+        /**
+         * Returns the base web socket uri
+         *
+         * @return the base web socket uri
+         */
+        URI getUri();
+
+        /**
+         * Returns the object mapper used for (de)serialisation of websocket messages
+         *
+         * @return the object mapper used for (de)serialisation of websocket messages
+         */
+        ObjectMapper getObjectMapper();
+
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(WebSocketClient.class);
+
+    private final String token;
+    private final URI uri;
+    private final ObjectMapper objectMapper;
+    private final BlockingQueue<WebSocketMessage> next = new LinkedBlockingQueue<>(1);
```

Comment:
This is a depth of 1 as it's the handshake between the WebSocket and stream. The stream's emitter could be passed into the WebSocketListener as then as soon as a message arrives over the socket it would be emitted into the stream. As an RxJava Flowable is being used that provides back pressure, the messages will be Buffered. But the problem there is that there is a possibility for messages to be lost. This was the best solution I could see to this. The best solution is to only send a CTS message back when the message has been consumed on the client. We could chalk that up in another issue?